hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/HeapData.java

```diff
@@ -63,14 +64,19 @@ public int totalSize() {
     @Override
     public int getPartitionHash() {
         if (hasPartitionHash()) {
-            return Bits.readIntB(payload, payload.length - Bits.INT_SIZE_IN_BYTES);
+            return Bits.readIntB(payload, payload.length - INT_SIZE_IN_BYTES);
         }
         return hashCode();
     }
 
     @Override
     public boolean hasPartitionHash() {
-        return totalSize() != 0 && payload[PARTITION_HASH_BIT_OFFSET] != 0;
+        int i = 1;
+        boolean payloadCheck = payload != null && payload.length > INT_SIZE_IN_BYTES;
+        return payloadCheck && payload[payload.length - i++] != 0
+                && payload[payload.length - i++] != 0
```

Comment:
What if the first byte of the partition hash is 0? It will short circuit and return false.